### PR TITLE
executor: reduce memory usage and GC overhead for hash join.

### PR DESCRIPTION
**Pull request by coocood**
Implemented and use MVMap to reduce GC overhead and memory usage for hash join.

### CODE REVIEW

**Comment by coocood** [comment]
benchmark result will be added later.

File: executor/join.go [hunk 1]
```diff
@@ -231,18 +237,19 @@ func (e *HashJoinExec) prepare() error {
...
 			return errors.Trace(err)
 		}
 		if hasNull {
 			continue
 		}
-		if rows, ok := e.hashTable[string(hashcode)]; !ok {
-			e.hashTable[string(hashcode)] = []*Row{row}
-		} else {
-			e.hashTable[string(hashcode)] = append(rows, row)
+		buffer = buffer[:0]
```
Comment (hanfei1991) [inline executor/join.go]: Why do this ?

File: executor/join.go [hunk 2]
```diff
@@ -257,6 +264,78 @@ func (e *HashJoinExec) prepare() error {
...
+	b = codec.EncodeVarint(b, int64(len(row.RowKeys)))
+	for _, rowKey := range row.RowKeys {
+		// Every rowKey is encoded with handle and idx in tableNames.
+		b = e.encodeRowKey(b, rowKey)
+	}
+	b, err := codec.EncodeValue(b, row.Data...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return b, nil
```
Comment (hanfei1991) [inline executor/join.go]: return b, errors.Trace(err)

File: executor/join.go [hunk 3]
```diff
@@ -257,6 +264,78 @@ func (e *HashJoinExec) prepare() error {
...
+}
+
+func (e *HashJoinExec) encodeRowKey(b []byte, rowKey *RowKeyEntry) []byte {
+	b = codec.EncodeVarint(b, rowKey.Handle)
+	for i, tn := range e.tableNames {
```
Comment (hanfei1991) [inline executor/join.go]: Every row's table name is the same. Need we encode it?

File: util/mvmap/mvmap.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+// NewMVMap creates a new multi-value map.
+func NewMVMap() *MVMap {
+	m := new(MVMap)
+	m.hashTable = make(map[uint64]entryAddr)
+	m.hashFunc = fnv.New64()
+	m.entryStore.slices = [][]entry{make([]entry, 0, 64)}
+	// append first empty entry so zero entry pointer an represent null.
+	m.entryStore.put(entry{})
```
Comment (hanfei1991) [inline util/mvmap/mvmap.go]: put `nullEntryAddr`

File: util/mvmap/mvmap.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+const (
+	maxDataSliceLen  = 64 * 1024
+	maxEntrySliceLen = 8 * 1024
+)
+
+func (ds *dataStore) put(key, value []byte) dataAddr {
+	dataLen := uint32(len(key) + len(value))
+	if ds.sliceLen != 0 && ds.sliceLen+dataLen > maxDataSliceLen {
+		ds.slices = append(ds.slices, make([]byte, 0, maxDataSliceLen))
```
Comment (hanfei1991) [inline util/mvmap/mvmap.go]: the cap should be `max(maxDataSliceLen, dataLen)` ?

File: util/mvmap/mvmap.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	m.hashTable[hashKey] = newEntryPtr
+}
+
+// Get gets the values of the key.
+func (m *MVMap) Get(key []byte) [][]byte {
+	var values [][]byte
+	hashKey := m.hash(key)
+	entryAddr := m.hashTable[hashKey]
+	for entryAddr != nullEntryAddr {
+		he := m.entryStore.get(entryAddr)
```
Comment (hanfei1991) [inline util/mvmap/mvmap.go]: Why name he? It makes me confused.

File: util/mvmap/mvmap.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		entryAddr = he.next
+		k, v := m.dataStore.get(he)
```
Comment (hanfei1991) [inline util/mvmap/mvmap.go]: can we implement compare and get ? If key is different with k, don't fetch the v.

File: executor/join.go (hunk 1 above)
Comment (coocood) [inline executor/join.go]: Reuse the buffer memory.

File: util/mvmap/mvmap.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func (es *entryStore) get(addr entryAddr) entry {
+	return es.slices[addr.sliceIdx][addr.offset]
+}
+
+// MVMap stores multiple value for a given key with minimum GC overhead.
+// A given key can store multiple values.
+// It is not thread-safe, should only be used in one goroutine.
+type MVMap struct {
```
Comment (tiancaiamao) [inline util/mvmap/mvmap.go]: This struct is complex, what's the benefit versus this?
```
type KeyType []byte
type ValueType [][]byte
map[KeyType]ValueType
```

File: executor/join.go [hunk 9]
```diff
@@ -257,6 +264,78 @@ func (e *HashJoinExec) prepare() error {
...
+	values := make([]types.Datum, e.smallExec.Schema().Len())
+	err = codec.SetRawValues(data, values)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	err = decodeRawValues(values, e.smallExec.Schema())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	row.Data = values
```
Comment (tiancaiamao) [inline executor/join.go]: I'm confused with Row struct, it has a Data and a RowKeys fields, what's the difference?

**Comment by tiancaiamao** [comment]
Does the encode/decode overhead cover the GC cost after change Row to []byte ?

**Comment by coocood** [comment]
Tested join two tables with 1 million rows, the TiDB process memory has increased to 4GB, a little bit smaller than master, this probably due to the GC is slower than memory allocation.

```
./benchdb -run="create|insert:0_1000000"
./benchdb -table="benchdb2" -run="create|insert:0_1000000"
```
```
select * from benchdb, benchdb2 where benchdb.id = benchdb2.id and benchdb.exp + benchdb2.exp < 0;
```

The performance is about 10% faster than master when hash table size is 1 million, but slower for smaller hash table size.

I'll try to implement lazy decode on hash join, and see the difference.

File: util/mvmap/mvmap.go (hunk 8 above)
Comment (coocood) [inline util/mvmap/mvmap.go]: If the key type and value type in the map don't have pointer, Go garbage collector simply skip the map.

**Comment by hanfei1991** [comment]
LGTM

**Comment by coocood** [comment]
After remove the extra memory allocation.
The benchmark result is:

memory:
```
master: 3.3GB
encode-row: 3.0GB
```

time:
```
master: 7.6s            
encode-row: 6.8s
```

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**